### ISSUE TITLE
assert size/strides for fallback kernel

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#144717]
### 🐛 Describe the bug

Inductor right now does not generate size/stride asserts for fallback kernel. This makes issue like [this](https://fb.workplace.com/groups/1075192433118967/posts/1567334737238065) very hard to debug (this is a meta internal link).

Actually in ir.FallbackKernel, we have the following code whose intention is to assert for size/strides for fallback kernel:

https://github.com/pytorch/pytorch/blob/c15d6508bdb82580803ea4899230043bf6ac2c04/torch/_inductor/ir.py#L6669-L6670

However, Fallback kernel usually generate a node with MultiOutputLayout which does not pass the if check.

A fix is to iterate thru each item for the FallbackKernel (check self.outputs) and assert size/stride for each of them.

I use the following testing script:
```
import torch
import einops
from torch._inductor import config as inductor_config
from torch._dynamo.testing import rand_strided, reset_rng_state

inductor_config.fallback_random = True

image_latent = torch.randn((24, 16, 32, 32), device="cuda").to(memory_format=torch.channels_last).view(2, 12, 16, 32, 32)

def f(image_latent):
    indices = torch.argsort(torch.rand(2, 12), dim=-1)[:, : 6]

    tar_latent = image_latent[
        torch.arange(2).unsqueeze(-1), indices[:, 3:]
    ]

    tar_latent_rearranged = einops.rearrange(
        tar_latent, "b n c h w -> (b n) c h w"
    )
    return {
        "tar_latent": tar_latent,
        "tar_latent_rearranged": tar_latent_rearranged,
    }


reset_rng_state()
ref = f(image_latent)
opt_f = torch.compile(f)
reset_rng_state()
act = opt_f(image_latent)
print(f"max dif {(act['tar_latent'] - ref['tar_latent']).abs().max()}")
print(f"max dif {(act['tar_latent_rearranged'] - ref['tar_latent_rearranged']).abs().max()}")
```

The script may not be able to repro anymore once we fix the layout problem for index.Tensor .

### Versions

..

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @muchulee8 @ColinPeppler @amjames @desertfire @aakhundov @BoyuanFeng @eellison

**Comment by ezyang** [comment]
Here's another example where lack of stride asserts on fallback is problematic:

Steps to reproduce:
1. Patch in https://github.com/pytorch/pytorch/pull/145080 as of 6a2829e6161cc6f135431c4de270e9e5f0838043
2. Run `PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=3 python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCUDA.test_comprehensive_fft_irfft2_cuda_float32`

This test fails accuracy, almost certainly because the stride metas at time of writing are wrong (the metas for fft are written w.r.t. CPU MKL, and not the CUDA implementation). But it just silently produces wrong accuracy.

**Comment by shunting314** [comment]
I ran ```PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=3 python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCUDA.test_comprehensive_fft_irfft2_cuda_float32``` with PR https://github.com/pytorch/pytorch/pull/145904 and see size/strides asserts in the generated code: https://gist.github.com/shunting314/6c429a6a069221c9bb53998233b30e36 . Didn't patch since the PR is already in.
